### PR TITLE
refactor autograd_hook

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54995 static graph api skeleton
* **#54981 refactor autograd_hook**
* #56730 add a test for grad view with torch amp

put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

### CODE REVIEW

**Review by wayi1 — APPROVED** [review]
Thanks for the refactoring!

File: torch/lib/c10d/reducer.hpp [hunk 1]
```diff
@@ -422,7 +419,11 @@ class Reducer {
  private:
   void reset_bucket_counting();
   void search_unused_parameters(
-    const std::vector<torch::autograd::Variable>& outputs);
+      const std::vector<torch::autograd::Variable>& outputs);
+  void set_divide_factor();
```
Comment (rohan-varma) [inline torch/lib/c10d/reducer.hpp:423]: nit: would be useful to add comments for these functions

**Review by rohan-varma — APPROVED** [review]
Thanks! It's great that we're making reducer code more modular and readable.

File: torch/lib/c10d/reducer.cpp [hunk 2]
```diff
@@ -421,6 +438,14 @@ void Reducer::autograd_hook(VariableIndex index) {
   // Carry over thread local state from main thread. This allows for
   // thread-local flags such as profiler enabled to be configure correctly.
   at::ThreadLocalStateGuard g(thread_local_state_);
+
+  // Ignore if we don't expect to be called.
+  // This may be the case if the user wants to accumulate gradients
+  // for number of iterations before reducing them.
+  if (!expect_autograd_hooks_) {
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/lib/c10d/reducer.cpp:445]: I think it is the right place, if sync is not needed, local_used_maps_ does not need to be set as well.

**Pull request closed without merging**